packages/adminrouter/extra/src/README.md

```diff
@@ -716,6 +716,222 @@ done by `test-harness/modules/mocker/jwt.py` module, which together with
 `repo_is_ee` fixture provides abstracts away the type of the
 token used by Admin Router itself.
 
+### Tests code reuse and deduplication
+Lots of endpoints exposed by Admin Router share common behaviour. Usually, it
+is something simple like testing for correct headers, the right upstream, etc...
+This results in tests covering them being very repetitive.
+
+In order to keep the code DRY, common code was extracted into a library with
+generic tests and helpers and a testing framework using YAML configuration files
+was developed.
+
+#### Generic tests
+Tests shared by most of the endpoints are as follows:
+* `generic_no_slash_redirect_test` - test that a location without a trailing
+  slash is redirected to one that ends with `/`.
+* `generic_response_headers_verify_test` - test that the response sent by AR is
+  correct, this mostly involves checking response headers.
+* `generic_upstream_headers_verify_test` - test that the request sent by NGINX to
+  the upstream is correct. Again - this is mostly about the headers now.
+* `generic_correct_upstream_dest_test` - test that the upstream request done by
+  Admin Router is sent to the correct upstream.
+* `generic_correct_upstream_request_test` - test that the path component of the
+  request sent to the upstream by AR, along with HTTP version, is correct.
+* `generic_location_header_during_redirect_is_adjusted_test` - test that
+  `Location` headers from the response sent by the upstream are being
+  rewritten/adjusted by NGINX before sending them to the client.
+
+All of these tests can be found in `test-harness/modules/generic_test_code/\*.py`
+together with extensive docstrings which explain all the call parameters.
+
+#### Helpers
+Helpers shared by most of the tests are as follow:
+* `overridden_file_content` - temporarily, while in the scope of the context
+  manager, substitute the contents of the given file with a given content. At
+  the end of the context, the original file contents are restored.
+* `verify_header` and `header_is_absent` - these functions check if the given
+  list of header tuples contains or does not contain given header.
+* `GuardedSubprocess` - launches given AR instance and makes sure that it is
+  terminated at the end of the context.
+* `iam_denies_all_requests` - makes sure that the IAM denies all PolicyQuery
+  requests within the context.
+* `ping_mesos_agent` - tests if agent with a given ID
+  (`mocker.endpoints.mesos.AGENT1_ID` by default) can be reached.
+* `repo_is_ee` - detects repository flavour - whether repository is Open DC/OS
+  or Enterprise DC/OS.
+* `assert_iam_queried_for_uid` - asserts that within the context, IAM has been
```

Comment:
In the big PR I asked a question about this one, and was hoping for clarification.

Did you consider my comments in the big PR?